File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
#include <algorithm>

using namespace std;

void bfs(int n, vector<vector<int>> &graph, vector<int> &visited) {
    queue<pair<int, int>> q; // {node, res}
    q.push({1, 0});
    visited[1] = -1;
    
    while (!q.empty()) {
        int node = q.front().first;
        int res = q.front().second;
        q.pop();
        
        for (int i = 0; i < graph[node].size(); i++) {
            if (visited[graph[node][i]] != 0) continue;
            
            visited[graph[node][i]] = res+1;
            q.push({graph[node][i], res+1});
        }
    }
}

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    vector<vector<int>> graph(n+1);
    vector<int> visited(n+1);
    int L = edge.size();
    
    for (int i = 0; i < L; i++) {
        graph[edge[i][0]].push_back(edge[i][1]);
        graph[edge[i][1]].push_back(edge[i][0]);
    }
    
    // for (int i = 1; i <= n; i++) {
    //     for (int j = 0; j < graph[i].size(); j++) cout << graph[i][j] << " ";
    //     cout << "\n";
    // }
    
    bfs(n, graph, visited);
    int target = *max_element(visited.begin(), visited.end());
    for (auto v : visited) if (target == v) answer++;
    return answer;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
import java.io.*;

class Solution {
    static int answer = 0, maxLength = 0;
    static ArrayList<ArrayList<Integer>> graph;
    
    static void bfs(int start, int[] res, int n) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);
        res[start] = 0;
        
        while (!q.isEmpty()) {
            int node = q.poll();
            
            for (int nNode : graph.get(node)) {
                if (res[nNode] == 0 && nNode != start) {
                    res[nNode] = res[node]+1;
                    q.offer(nNode);
                }
            }
        }
        
        for (int i = 1; i <= n; i++) maxLength = Math.max(maxLength, res[i]);
        
        for (int i = 1; i <= n; i++) if (maxLength == res[i]) answer++;
    }
    
    public int solution(int n, int[][] edge) {
        boolean[] visited = new boolean[n+1];
        Arrays.fill(visited, false);
        graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());
        
        for (int i = 0; i < edge.length; i++) {
            int a = edge[i][0];
            int b = edge[i][1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        
        int[] res = new int[n+1];
        Arrays.fill(res, 0);
        bfs(1, res, n);
        
        return answer;
    }
}